crates/forge/src/lib.rs

```diff
@@ -435,14 +408,67 @@ fn run_tests_from_crate(
     ))
 }
 
-fn run_with_fuzzing(
-    package_root: &Utf8PathBuf,
-    runner_config: &RunnerConfig,
-    runner_params: &RunnerParams,
-    runner: &SierraCasmRunner,
-    case: &TestCase,
-    args: &Vec<&ConcreteTypeId>,
-) -> Result<(TestCaseSummary, u32)> {
+fn choose_test_strategy_and_run(
+    case: Arc<TestCase>,
+    runner: Arc<SierraCasmRunner>,
+    runner_config: Arc<RunnerConfig>,
+    args: Vec<ConcreteTypeId>,
+) -> JoinHandle<Result<TestCaseSummary>> {
+    if args.is_empty() {
+        run_single_test(case, runner, runner_config)
+    } else {
+        tokio::task::spawn(async move {
+            run_with_fuzzing(runner_config.clone(), runner, case, &args).await
+        })
+    }
+}
+
+fn run_single_test(
+    case: Arc<TestCase>,
+    runner: Arc<SierraCasmRunner>,
+    runner_config: Arc<RunnerConfig>,
+) -> JoinHandle<Result<TestCaseSummary>> {
+    let exit_first = runner_config.exit_first;
+    tokio::task::spawn(async move {
+        tokio::select! {
+            () = runner_config.exit_first_cancellation_token.cancelled() => {
+                //Stop executing all tests because flag --exit-first'
+                //has been set and one test FAIL
+                Ok(TestCaseSummary::skipped(&case))
+            },
+            () = runner_config.error_cancellation_token.cancelled() => {
+                //Stop executing all tests because
+                //one of a test returns Err
+                Ok(TestCaseSummary::None {  })
+            },
+            result = blocking_run_from_test(runner, case.clone(), runner_config.clone(), vec![], None) => {
+                match result {
+                    Ok(result) => {
+                        if exit_first {
+                            if let TestCaseSummary::Failed { .. } = &result {
+                                runner_config.exit_first_cancellation_token.cancel();
+                            }
+                        }
+                        Ok(result)
+                    }
+                    Err(e) => {
+                        runner_config.error_cancellation_token.cancel();
+                        Err(e)
+                    }
+                }
+            }
+        }
+    })
+}
+
+async fn run_with_fuzzing(
+    runner_config: Arc<RunnerConfig>,
+    runner: Arc<SierraCasmRunner>,
+    case: Arc<TestCase>,
```

Comment:
I don't think this needs to be `Arc`, we can just consume the test case in the task.

Reply:
`run_from_test_case` get `case` as params, and in case of fuzzing we create in default 256 task which use `case`